### PR TITLE
GenericHub and GenericRepeater Components

### PR DESCRIPTION
**Pull request by LeStarch**
| | |
|:---|:---|
|**_Originating Project/Creator_**| Infrastructure |
|**_Affected Component_**| Svc::GenericHub and Svc::GenericRepeater  |
|**_Affected Architectures(s)_**|  |
|**_Related Issue(s)_**|  |
|**_Has Unit Tests (y/n)_**| y |
|**_Builds Without Errors (y/n)_**| y |
|**_Unit Tests Pass (y/n)_**|y |
|**_Documentation Included (y/n)_**|  y|

---
## Change Description

This adds 2 components, a Svc::GenericHub and Svcx::GenericRepeater.

## Rationale

These are easily sharable common components.

## Testing/Review Recommendations

Reviewers should ignore the Tcp rework, as that is captured in: #316.  This was required to be built on-top of that rework.

## Future Work

None.

### CODE REVIEW

**Review by kevin-f-ortega — COMMENTED** [review]
Looks good overall. I added few comments.

File: Svc/GenericRepeater/GenericRepeaterComponentImpl.cpp [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+  void GenericRepeaterComponentImpl ::
+    portIn_handler(
+        NATIVE_INT_TYPE portNum, /*!< The port number*/
+        Fw::SerializeBufferBase &Buffer /*!< The serialization buffer*/
+    )
+  {
+    for(NATIVE_INT_TYPE i = 0; i < NUM_PORTOUT_OUTPUT_PORTS; i++) {
+      if(!isConnected_portOut_OutputPort(i)) {
+        break;
```
[the remark this answers is not in the document]
Comment (LeStarch) [reply Svc/GenericRepeater/GenericRepeaterComponentImpl.cpp:57]: I believe this was the point.  Assuming only the lowest ports are connected, then this prevents needless iterating.  Is it a common pattern to use this with interleaved connections?

[the remark this answers is not in the document]
Comment (LeStarch) [reply Svc/GenericRepeater/GenericRepeaterComponentImpl.cpp:57]: I suppose so, huh?

File: Svc/CMakeLists.txt [hunk 2]
```diff
@@ -34,3 +34,5 @@ add_fprime_subdirectory("${CMAKE_CURRENT_LIST_DIR}/PrmDb/")
 add_fprime_subdirectory("${CMAKE_CURRENT_LIST_DIR}/RateGroupDriver/")
 add_fprime_subdirectory("${CMAKE_CURRENT_LIST_DIR}/Time/")
 add_fprime_subdirectory("${CMAKE_CURRENT_LIST_DIR}/TlmChan/")
+add_fprime_subdirectory("${CMAKE_CURRENT_LIST_DIR}/GenericHub/")
```
[the remark this answers is not in the document]
Comment (LeStarch) [reply Svc/CMakeLists.txt]: Should be fixed.

File: Svc/GenericHub/GenericHubComponentImpl.cpp [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+  GenericHubComponentImpl ::
+    ~GenericHubComponentImpl(void)
+  {
+
+  }
+
+  void GenericHubComponentImpl ::
+    send_data(const HubType type, const NATIVE_INT_TYPE port, const U8* data, const U32 size)
+  {
+      Fw::SerializeStatus status;
```
[the remark this answers is not in the document]
Comment (LeStarch) [reply Svc/GenericHub/GenericHubComponentImpl.cpp]: Should be fixed.

File: Svc/GenericHub/GenericHubComponentImpl.cpp [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+  // ----------------------------------------------------------------------
+  // Construction, initialization, and destruction
+  // ----------------------------------------------------------------------
+
+  GenericHubComponentImpl ::
+    GenericHubComponentImpl(
+        const char *const compName
+    ) : GenericHubComponentBase(compName)
+  {
+
```
[the remark this answers is not in the document]
Comment (LeStarch) [reply Svc/GenericHub/GenericHubComponentImpl.cpp:27]: Should be fixed.

File: Svc/GenericRepeater/GenericRepeaterComponentImpl.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (LeStarch) [reply Svc/GenericRepeater/GenericRepeaterComponentImpl.cpp]: Should be fixed.

File: Svc/GenericHub/GenericHubComponentImpl.hpp [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+       * Type of serialized data on the wire. Allows for expanding them on the opposing end.
+       */
+      enum HubType {
+          HUB_TYPE_PORT, //!< Port type transmission
+          HUB_TYPE_BUFFER, //!< Buffer type transmission
+          HUB_TYPE_MAX
+      };
+
+
+      const static U32 GENERIC_HUB_DATA_SIZE = 1024;
```
[the remark this answers is not in the document]
Comment (LeStarch) [reply Svc/GenericHub/GenericHubComponentImpl.hpp]: Once we have a better Buffer manager, we may make this allocate buffers on the fly.

[the remark this answers is not in the document]
Comment (LeStarch) [reply Svc/GenericHub/GenericHubComponentImpl.hpp]: So for now I will leave this.

File: Svc/GenericHub/GenericHubComponentImpl.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (LeStarch) [reply Svc/GenericHub/GenericHubComponentImpl.cpp]: I remove the size > 0 check.  The buffer should never be null, but the hub works correctly on zero sized data and port calls without args may pass zero sized data.

File: Svc/GenericRepeater/GenericRepeaterComponentImpl.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (kevin-f-ortega) [reply Svc/GenericRepeater/GenericRepeaterComponentImpl.cpp]: > I believe this was the point. Assuming only the lowest ports are connected, then this prevents needless iterating. Is it a common pattern to use this with interleaved connections?

I do not see it a common pattern. But I could see the case when a topology connects ports 0, 1, 2, and 3. Then later on want to remove port connection 1. The hub would stop working and the user would be left scratching their head for a bit.

**Review by kevin-f-ortega — APPROVED** [review]
LGTM.

File: Svc/GenericRepeater/GenericRepeaterComponentImpl.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (timcanham) [reply Svc/GenericRepeater/GenericRepeaterComponentImpl.cpp]: I agree that we shouldn't quit early. It isn't intuitive and could cause confusion.

**Comment by LeStarch** [comment]
Depends on #316 before merge.

**Comment by LeStarch** [comment]
Note: we are reworking these along with a more modular ground system.  Expect new PRs in the near fututre.

**Pull request closed without merging**